### PR TITLE
70 token cache refresh

### LINKED ISSUE #70: Use a percentage-based strategy to refresh cached tokens
In the `TokenCache` we currently refresh tokens based on a hardcoded threshold: https://github.com/zalando-incubator/lib-oauth-tooling/blob/master/src/TokenCache.ts#L8, used in https://github.com/zalando-incubator/lib-oauth-tooling/blob/master/src/TokenCache.ts#L96

To be more resilient we should refresh tokens based on the percentage of their lifetime, for example when 50% of the expiration time is met. We should default this value to 50% and provide an option to be able to parameterise this value.

### PR DESCRIPTION
**Pull request by bzums**
close #70

### CODE REVIEW

File: src/TokenCache.ts [hunk 1]
```diff
@@ -78,14 +83,17 @@ class TokenCache {
         };
 
         return getAccessToken(config)
-          .then((token) => getTokenInfo(this.oauthConfig.tokenInfoEndpoint, token.access_token))
           .then((token) => {
 
-            token.local_expiry = Date.now() + token.expires_in * 1000 - EXPIRE_THRESHOLD;
-            this._tokens[tokenName] = token;
+            const localExpiry = Date.now() + (token.expires_in * 1000 * DEFAULT_PERCENTAGE_LEFT);
```
Comment (bzums) [inline src/TokenCache.ts]: should be `const localExpiry = Date.now() + (token.expires_in * 1000 * (1 - DEFAULT_PERCENTAGE_LEFT));`

**Comment by Retro64** [comment]
:+1:

File: integration-test/oauth-tooling/tokenCache.spec.ts [hunk 2]
```diff
@@ -113,111 +109,44 @@ describe('tokenCache', () => {
...
+    return expect(promise).to.become(defaultAccessTokenValue);
   });
 
   it('#get should resolve with a new token if the cached one is expired', () => {
 
     // given
-    const firstAccessToken = '4b70510f-be1d-4f0f-b4cb-edbca2c79d41';
-    const secondAccessToken = '9sdf8fd8-be1d-4f0f-b4cb-54nk66n45knk';
+    const clock = lolex.install();
+    const initialLifetime = 3600;
```
Comment (ISO50) [inline integration-test/oauth-tooling/tokenCache.spec.ts]: // in ms?

File: package.json [hunk 3]
```diff
@@ -42,6 +43,7 @@
     "@types/chai-as-promised": "0.0.30",
     "@types/express": "4.0.35",
     "@types/http-status": "0.2.29",
+    "@types/lolex": "^1.5.31",
```
Comment (ISO50) [inline package.json]: `^`

**Comment by bzums** [comment]
👍

File: src/TokenCache.ts [hunk 4]
```diff
@@ -78,14 +83,17 @@ class TokenCache {
...
+            const localExpiry = Date.now() + (token.expires_in * 1000 * (1 - DEFAULT_PERCENTAGE_LEFT));
+            this._tokens[tokenName] = {
+              ...token,
+              local_expiry: localExpiry
+            };
 
             return token;
-          });
+          })
+          .then((token) => getTokenInfo(this.oauthConfig.tokenInfoEndpoint, token.access_token));
```
Comment (bzums) [inline src/TokenCache.ts]: This tokeninfo call may be unnecessary.

**Comment by bzums** [comment]
👍

**Comment by Retro64** [comment]
:+1:

**Pull request merged**